lib/libzfs/libzfs_changelist.c

```diff
@@ -409,11 +409,9 @@ change_one(zfs_handle_t *zhp, void *data)
 	 * order to do the rename.  Also, if this is a volume and we're doing
 	 * a rename, then always add it to the changelist.
 	 */
-
-	if (!(ZFS_IS_VOLUME(zhp) && clp->cl_realprop == ZFS_PROP_NAME) &&
-	    zfs_prop_get(zhp, clp->cl_prop, property,
-	    sizeof (property), &sourcetype, where, sizeof (where),
-	    B_FALSE) != 0) {
+	if (zfs_prop_get(zhp, clp->cl_prop, property, sizeof (property),
+	    &sourcetype, where, sizeof (where), B_FALSE) != 0 &&
+	    !(ZFS_IS_VOLUME(zhp) && clp->cl_realprop == ZFS_PROP_NAME)) {
```

Comment:
@behlendorf  At first glace, only initializing `sourcetype` to `ZPROP_SRC_NONE` instead of reordering these, we may lost the dataset in the changelist when we rename volume: it won't call `zfs_prop_get` actually in this case and `sourcetype` will keeps the default value of `ZPROP_SRC_NONE`. 
But, `clp->cl_alldependents` always set to `B_TRUE` in rename case and we will add the dataset to changelist no matter what's the value of `sourcetype`, that's seems OK then.
Thanks for your suggestion!

```
	if (!(ZFS_IS_VOLUME(zhp) && clp->cl_realprop == ZFS_PROP_NAME) &&
	    zfs_prop_get(zhp, clp->cl_prop, property,
	    sizeof (property), &sourcetype, where, sizeof (where),
	    B_FALSE) != 0) {
```